File: haskell/pr4.hs
-- A palindromic number reads the same both ways. The largest
-- palindrome made from the product of two 2-digit numbers is 9009 =
-- 91 × 99.
--
-- Find the largest palindrome made from the product of two 3-digit
-- numbers.

module Main where

import Data.List
import Data.Ord
import Nums

main :: IO ()
main = putStrLn $ show solution

isPalindrome :: Eq a => [a] -> Bool
isPalindrome x = x == reverse x

solution :: (Int, Int, Int)
solution = maximumBy (comparing (\ (_, _, x) -> x)) pals

pals :: [(Int, Int, Int)]
pals = [ (x, y, prod) | x <- [100..999], y <- [x..999],
         let prod = x * y,
         isPalindrome $ digits prod ]

File: rust/pr4.rs
// Problem 4
// 
// 16 November 2001
// 
// A palindromic number reads the same both ways. The largest palindrome made
// from the product of two 2-digit numbers is 9009 = 91 × 99.
// 
// Find the largest palindrome made from the product of two 3-digit numbers.

use std;

fn main() {
    let max = 0u;
    uint::range(100u, 1000u) { |a|
        uint::range(a, 1000u) { |b|
            let c = a * b;
            if c > max && is_palindrome(c) {
                max = c;
            }
        }
    }
    std::io::println(#fmt("%u", max));
}

fn is_palindrome(n: uint) -> bool {
    n == reverse_number(n)
}

fn reverse_number(number: uint) -> uint {
    let n = number;
    let result = 0u;
    while n > 0u {
        result = result * 10u + n % 10u;
        n /= 10u;
    }
    result
}
